# Practical 9b: Gradient Descent Algorithm

To obtain a local minimum of a function $z=f(x,y)$, the gradient descent algorithm can be implemented. The partial derivatives of the function, $\frac{\partial z}{\partial x}$ and $\frac{\partial z}{\partial y}$ are needed and can be obtained using the SymPy library. The algorithm works this way,

Step 0.	Set a learning rate $\alpha>0$ and an initial point $x=x_0,\ y=y_0$ and compute $f(x_0,y_0)$.

Step 1.	At n-th point $x=x_n,\ y=y_n$, compute ${{f}_{x}}({{x}_{n}},{{y}_{n}})$ and $f_y(x_n,y_n)$.

Step 2.	Update to the (n+1)-th point, $x_{n+1}=x_n-\alpha f_x(x_n,y_n), \ y_{n+1}=y_n-\alpha f_y(x_n,y_n)$ and compute $f(x_{n+1},y_{n+1})$. 

Step 3.	Repeat step 1 and 2 until a stopping criterion is reached.

Each or a combination of the following can be the stopping criterion:    
1.	The **maximum number of iterations** is reached.
2.	The **value of $f_x^2(x_n,y_n)+ f_y^2(x_n,y_n)$** is smaller than a fixed constant. 
3.	**Convergence**, which, in simple terms, means that the update to the current point does not differ much from the previous point. It can refer to little difference from $x=x_n,\ y=y_n$ to $x=x_{n+1},\ y=y_{n+1}$ or little reduction in $f(x_{n},y_{n})$ to $f(x_{n+1},y_{n+1})$ (the difference is smaller than a fixed constant).
 
As an example, we want to find the minimum of the function $f(x,y)=x^2+3y^2$. The partial derivatives are $f_x(x,y) = 2x$ and $f_y(x,y)=6y$. With an initial point of $x=4,\ y=5$, learning rate $\alpha=0.1$, the code for the algorithm is as follows,

In [2]:
import numpy as np

next_x = 4 # Initial point
next_y = 5 # Initial point
alpha = 0.5 # Learning rate
epsilon = 0.001 # Stopping criterion constant
max_iters = 500 # Maximum number of iterations

# Partial derivatives and function
partialf_x = lambda x,y: 2*x 
partialf_y = lambda x,y: 6*y 
func = lambda x,y: x**2+3*y**2

next_func = func(next_x,next_y) # Initial value of function

for n in range(max_iters):
    current_x = next_x
    current_y = next_y
    current_func = next_func
    next_x = current_x-alpha*partialf_x(current_x,current_y) # update of x
    next_y = current_y-alpha*partialf_y(current_x,current_y) # update of y
    next_func = func(next_x,next_y)
    change_func = abs(next_func-current_func) # stopping criterion: values of function converge
    print("Iteration",n+1,": x = ",next_x,", y = ",next_y,", f(x,y) = ",next_func)
    if change_func<epsilon:
        break

Iteration 1 : x =  0.0 , y =  -10.0 , f(x,y) =  300.0
Iteration 2 : x =  0.0 , y =  20.0 , f(x,y) =  1200.0
Iteration 3 : x =  0.0 , y =  -40.0 , f(x,y) =  4800.0
Iteration 4 : x =  0.0 , y =  80.0 , f(x,y) =  19200.0
Iteration 5 : x =  0.0 , y =  -160.0 , f(x,y) =  76800.0
Iteration 6 : x =  0.0 , y =  320.0 , f(x,y) =  307200.0
Iteration 7 : x =  0.0 , y =  -640.0 , f(x,y) =  1228800.0
Iteration 8 : x =  0.0 , y =  1280.0 , f(x,y) =  4915200.0
Iteration 9 : x =  0.0 , y =  -2560.0 , f(x,y) =  19660800.0
Iteration 10 : x =  0.0 , y =  5120.0 , f(x,y) =  78643200.0
Iteration 11 : x =  0.0 , y =  -10240.0 , f(x,y) =  314572800.0
Iteration 12 : x =  0.0 , y =  20480.0 , f(x,y) =  1258291200.0
Iteration 13 : x =  0.0 , y =  -40960.0 , f(x,y) =  5033164800.0
Iteration 14 : x =  0.0 , y =  81920.0 , f(x,y) =  20132659200.0
Iteration 15 : x =  0.0 , y =  -163840.0 , f(x,y) =  80530636800.0
Iteration 16 : x =  0.0 , y =  327680.0 , f(x,y) =  322122547200.0
Iteration 17 : x =  0.0 , y =  -6553

The convergence criterion used in the above code is when the difference in the values of the function in 2 consecutive updates is less than a fixed constant, epsilon. The stopping criterion is the convergence criterion and when the maximum number of iterations is reached, whichever comes first.

Now, try to repeat the example above with different learning rates and observe what happens.

$\alpha=0.01$:  Converges, 161 iterations, x_min = 0.155, y_min = 0

$\alpha=0.05$:  Converges, 40 iterations, x_min = 0.059, y_min = 0 

$\alpha=0.1$: Converges, 21 iterations, x_min = 0.037 y_min = 0
    
$\alpha=0.2$: Converges, 11 iterations, x_min = 0.015, y_min = 0
    
$\alpha=0.25$: Converges, 10 iterations, x_min = 0.004, y_min = 0.005
    
$\alpha=0.5$: Diverges

Conclusion: Different learning rates yield different number of iterations. If the learning rate is too small, we may need a lot of iterations to reach the min value. If the learning rate is too large, we may end up not getting the minimum value (divergent).

Next, we want to look at how different initial points may result in different outputs. Given a function $f(x,y)=x^2+y^4+3y^3-y^2-3y$, we would like to find the local minimum using the gradient descent algorithm with different initial points. Fill in the code below with the appropriate parameters and functions.

In [5]:
# Feel free to set your own constants

import numpy as np

next_x = -1 # Initial point # students to set 
next_y = -2 # Initial point # students to set
alpha = 0.05 # Learning rate # students to set
epsilon = 0.001 # Stopping criterion constant # students to set
max_iters = 1000 # Maximum number of iterations # students to set

# Partial derivatives and function
partialf_x = lambda x,y: 2*x # students to fill in formula
partialf_y = lambda x,y: 4*y**3+9*y**2-2*y-3 # students to fill in formula
func = lambda x,y: x**2+y**4+3*y**3-y**2-3*y # students to fill in formula

next_func = func(next_x,next_y) # Initial value of function

for n in range(max_iters):
    current_x = next_x
    current_y = next_y
    current_func = next_func
    next_x = current_x-alpha*partialf_x(current_x,current_y) # update of x
    next_y = current_y-alpha*partialf_y(current_x,current_y) # update of y
    next_func = func(next_x,next_y)
    change_func = abs(next_func-current_func) # stopping criterion: values of function converge
    print("Iteration",n+1,": x = ",next_x,", y = ",next_y,", f(x,y) = ",next_func)
    if change_func<epsilon:
        break

Iteration 1 : x =  -0.9 , y =  -2.25 , f(x,y) =  -6.045468750000001
Iteration 2 : x =  -0.81 , y =  -2.325 , f(x,y) =  -6.257977734375005
Iteration 3 : x =  -0.7290000000000001 , y =  -2.3264156249999997 , f(x,y) =  -6.382655736907875
Iteration 4 : x =  -0.6561000000000001 , y =  -2.3263417107362816 , f(x,y) =  -6.483629578617903
Iteration 5 : x =  -0.5904900000000001 , y =  -2.3263456638058493 , f(x,y) =  -6.56541834866583
Iteration 6 : x =  -0.531441 , y =  -2.3263454526500595 , f(x,y) =  -6.631667252285251
Iteration 7 : x =  -0.4782969 , y =  -2.3263454639298313 , f(x,y) =  -6.685328864216634
Iteration 8 : x =  -0.43046721 , y =  -2.326345463327277 , f(x,y) =  -6.728794769881057
Iteration 9 : x =  -0.387420489 , y =  -2.326345463359465 , f(x,y) =  -6.764002153469246
Iteration 10 : x =  -0.3486784401 , y =  -2.326345463357746 , f(x,y) =  -6.792520134175677
Iteration 11 : x =  -0.31381059609 , y =  -2.3263454633578373 , f(x,y) =  -6.815619698547886
Iteration 12 : x =  -0.282429536481 

With all parameters constant except the initial points, what do you observe?

Initial point = (1,1): Converges, 26 iterations, x_min = 0.065, y_min = 0.607
    
Initial point = (2,1): Converges, 33 iterations, x_min = 0.062, y_min = 0.607
    
Initial point = (0,-3): Converges, 5 iterations, x_min = 0, y_min = -2.326
    
Initial point = (-1,-2): Converges, 26 iterations, x_min = -0.064, y_min = -2.326

Conclusion: Different initial points may yield different values for x and y. Sometimes you may end up with local optimum, and sometimes global optimum.
    
## Stopping criterion

So far, the stopping criterion has always been a combination of maximum number of iterations and convergence of the values of the function. Try editing the code to change the stopping criterion from the convergence of the function to the value of $f^2_x(x_n,y_n)+f^2_y(x_n,y_n)$ being smaller than a small number.

In [6]:
# Feel free to set your own constants

import numpy as np

next_x = 2 # Initial point # students to set 
next_y = 1 # Initial point # students to set
alpha = 0.05 # Learning rate # students to set
epsilon = 0.0001 # Stopping criterion constant # students to set
max_iters = 1000 # Maximum number of iterations # students to set

# Partial derivatives and function
partialf_x = lambda x,y: 2*x # students to fill in formula
partialf_y = lambda x,y: 4*y**3+9*y**2-2*y-3 # students to fill in formula
func = lambda x,y: x**2+y**4+3*y**3-y**2-3*y # students to fill in formula

next_func = func(next_x,next_y) # Initial value of function

for n in range(max_iters):
    current_x = next_x
    current_y = next_y
    current_func = next_func
    next_x = current_x-alpha*partialf_x(current_x,current_y) # update of x
    next_y = current_y-alpha*partialf_y(current_x,current_y) # update of y
    next_func = func(next_x,next_y)
    partial_norm = (partial_f(next_x,next_y)**2) + partial_f(next_x,next_y)**2 # stopping criterion: (f_x)^2+(f_y)^2 # student to fill in 
    print("Iteration",n+1,": x = ",next_x,", y = ",next_y,", f(x,y) = ",next_func)
    if partial_norm<epsilon:
        break

NameError: name 'partial_f' is not defined

With all parameters constant except the initial points, what do you observe?

epsilon = 0.01: Converges, 36 iterations, x_min = 0.045, y_min = 0.607, f=-1.38

epsilon = 0.001: Converges, 46 iterations, x_min = 0.016, y_min = 0.607, f=-1.38

epsilon = 0.0001: Converges, 57 iterations, x_min = 0.016, y_min = 0.607, f=-1.38

Conclusion: The smaller the epsilon, the more iterations it will take for the algorithm.